File: LeetCode/Python Solutions/Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/longest.py
from collections import deque
from typing import List

class Solution:
    def longestSubarray(self, nums: List[int], limit: int) -> int:
        maxDeque = deque()
        minDeque = deque()
        left = 0
        result = 0

        for right in range(len(nums)):
            while maxDeque and nums[maxDeque[-1]] <= nums[right]:
                maxDeque.pop()
            while minDeque and nums[minDeque[-1]] >= nums[right]:
                minDeque.pop()
            
            maxDeque.append(right)
            minDeque.append(right)
            
            while nums[maxDeque[0]] - nums[minDeque[0]] > limit:
                left += 1
                if maxDeque[0] < left:
                    maxDeque.popleft()
                if minDeque[0] < left:
                    minDeque.popleft()
            
            result = max(result, right - left + 1)
        
        return result

# Example usage:
sol = Solution()
print(sol.longestSubarray([8,2,4,7], 4))  # Output: 2
print(sol.longestSubarray([10,1,2,4,7,2], 5))  # Output: 4
print(sol.longestSubarray([4,2,2,2,4,4,2,2], 0))  # Output: 3

File: LeetCode/Java Solutions/Longest Continuous Subarray With Absolute Diff Less Than or Equal to Limit/longest.java
import java.util.Deque;
import java.util.LinkedList;

class Solution {
    public int longestSubarray(int[] nums, int limit) {
        Deque<Integer> maxDeque = new LinkedList<>();
        Deque<Integer> minDeque = new LinkedList<>();
        int left = 0;
        int result = 0;

        for (int right = 0; right < nums.length; right++) {
            while (!maxDeque.isEmpty() && nums[maxDeque.peekLast()] <= nums[right]) {
                maxDeque.pollLast();
            }
            while (!minDeque.isEmpty() && nums[minDeque.peekLast()] >= nums[right]) {
                minDeque.pollLast();
            }
            
            maxDeque.addLast(right);
            minDeque.addLast(right);
            
            while (nums[maxDeque.peekFirst()] - nums[minDeque.peekFirst()] > limit) {
                left++;
                if (maxDeque.peekFirst() < left) {
                    maxDeque.pollFirst();
                }
                if (minDeque.peekFirst() < left) {
                    minDeque.pollFirst();
                }
            }
            
            result = Math.max(result, right - left + 1);
        }
        
        return result;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.longestSubarray(new int[]{8,2,4,7}, 4));  // Output: 2
        System.out.println(sol.longestSubarray(new int[]{10,1,2,4,7,2}, 5));  // Output: 4
        System.out.println(sol.longestSubarray(new int[]{4,2,2,2,4,4,2,2}, 0));  // Output: 3
    }
}
